src/algebra/lie_algebra.lean

```diff
@@ -205,16 +209,191 @@ def of_associative_algebra (A : Type v) [ring A] [algebra R A] : lie_algebra R A
 An important class of Lie algebras are those arising from the associative algebra structure on
 module endomorphisms.
 -/
-def of_endomorphism_algebra (M : Type v)
+instance of_endomorphism_algebra (M : Type v)
   [add_comm_group M] [module R M] : lie_algebra R (module.End R M) :=
-of_associative_algebra R (module.End R M)
+of_associative_algebra (module.End R M)
+
+/--
+The adjoint action of a Lie algebra on itself.
+-/
+def Ad (x : L) : L →ₗ[R] L :=
+{ to_fun := has_bracket.bracket x,
+  add    := by { intros, apply lie_add },
+  smul   := by { intros, apply lie_smul } }
+
+set_option class.instance_max_depth 35
+lemma is_morphism_Ad (x y : L) : (Ad ⁅x, y⁆ : L →ₗ[R] L) = ⁅Ad x, Ad y⁆ :=
+begin
+  ext z,
+  suffices : ⁅⁅x, y⁆, z⁆ = ⁅x, ⁅y, z⁆⁆ + ⁅⁅x, z⁆, y⁆, by {
+    dunfold of_endomorphism_algebra of_associative_algebra Ad
+            has_bracket.bracket ring_commutator.commutator,
+    simpa using this,
+  },
+  rw [eq_comm, ←lie_skew ⁅x, y⁆ z, ←lie_skew ⁅x, z⁆ y, ←lie_skew x z, lie_neg, neg_neg,
+      ←sub_eq_zero_iff_eq, sub_neg_eq_add, lie_ring.jacobi],
+end
+
+/--
+The bracket of a Lie algebra as a bilinear map.
+-/
+def bil_lie : L →ₗ[R] L →ₗ[R] L :=
+{ to_fun := Ad,
+  add    := by { unfold Ad, intros, ext, simp [add_lie], },
+  smul   := by { unfold Ad, intros, ext, simp, } }
+
+/--
+The bracket of a Lie algebra as a morphism of Lie algebras.
+-/
+def bil_lie_morphism : L →ₗ⁅R⁆ module.End R L := { bracket := is_morphism_Ad, ..bil_lie }
 
 end lie_algebra
 
+section lie_subalgebra
+
+variables (R : Type u) (L : Type v) [comm_ring R] [add_comm_group L] [lie_algebra R L]
+
+/--
+A Lie subalgebra of a Lie algebra is submodule that is closed under the Lie bracket.
+This is a sufficient condition for the subset itself to form a Lie algebra.
+-/
+structure lie_subalgebra extends submodule R L :=
+(bracket : ∀ {x y}, x ∈ carrier → y ∈ carrier → ⁅x, y⁆ ∈ carrier)
```

Comment:
In line with `lie_add`, `lie_neg`, etc... and with `smul_mem` and `add_mem` etc..., this condition should be called `lie_mem` instead of `bracket`.